src/util/codec/mysql/json/json_modify.rs

```diff
@@ -0,0 +1,164 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::Json;
+use super::super::Result;
+use super::path_expr::{PathLeg, PathExpression};
+use std::mem;
+
+/// `ModifyType` is for modify a JSON.
+#[derive(Clone, Debug, PartialEq)]
+pub enum ModifyType {
+    /// `Insert` is for inserting a new element into a JSON.
+    Insert,
+    /// `Replace` is for replacing a old element from a JSON.
+    Replace,
+    /// `Set` = `Insert` | `Replace`
+    Set,
+}
+
+impl Json {
+    // Modifies a Json object by insert, replace or set.
+    // All path expressions cannot contain * or ** wildcard.
+    // If any error occurs, the input won't be changed.
+    pub fn modify(&mut self,
+                  path_expr_list: &[PathExpression],
+                  mut values: Vec<Json>,
+                  mt: ModifyType)
+                  -> Result<()> {
+        if path_expr_list.len() != values.len() {
+            return Err(box_err!("Incorrect parameter count"));
+        }
+        for expr in path_expr_list {
+            if expr.contains_any_asterisk() {
+                return Err(box_err!("Invalid path expression"));
+            }
+        }
+        for (expr, value) in path_expr_list.iter().zip(values.drain(..)) {
+            self.set_json(&expr.legs, value, &mt);
+        }
+        Ok(())
+    }
+
+    // `set_json` is used in Json::modify().
+    fn set_json(&mut self, path_legs: &[PathLeg], value: Json, mt: &ModifyType) {
+        if path_legs.is_empty() {
+            match *mt {
+                ModifyType::Replace | ModifyType::Set => {
+                    *self = value;
+                }
+                _ => {}
+            }
+            return;
+        }
+
+        let (current_leg, sub_path_legs) = (&path_legs[0], &path_legs[1..]);
+        let base_data = mem::replace(self, Json::None);
+
+        if let PathLeg::Index(i) = *current_leg {
+            let index = i as usize;
+            // If `j` is not an array, we should autowrap it to be an array.
+            // Then if the length of result array equals to 1, it's unwraped.
+            let (mut array, wrapped) = match base_data {
+                Json::Array(array) => (array, false),
+                _ => (vec![base_data], true),
+            };
+            if array.len() > index {
+                let mut chosen = &mut array[index];
+                chosen.set_json(sub_path_legs, value, mt);
+            } else if sub_path_legs.is_empty() && *mt != ModifyType::Replace {
+                // e.g. json_insert('[1, 2, 3]', '$[3]', "x") => '[1, 2, 3, "x"]'
+                array.push(value);
+            }
+            if (array.len() == 1) && wrapped {
+                *self = array.pop().unwrap();
+            } else {
+                *self = Json::Array(array);
+            }
+            return;
+        }
+
+        if let PathLeg::Key(ref key) = *current_leg {
+            if let Json::Object(mut map) = base_data {
```

Comment:
Maybe `if let Json::Object(ref mut map) = *self {}` is better.
and move above `mem::replace` into `PathLeg::Index(i) block?